server/src-rsr/catalog_metadata.sql

```diff
@@ -173,6 +173,15 @@ from
         where function_name = cc.function_name and function_schema = cc.function_schema
       ) fi on 'true'
   ) as computed_field,
+  (
+    select
+      json_build_object(
+        'custom_types',
+         coalesce((select custom_types from hdb_catalog.hdb_custom_types), '{}'),
+        'pg_scalars',
+         coalesce((select json_agg(typname) from pg_catalog.pg_type where typtype = 'b'), '[]')
+      ) as item
```

Comment:
This is the sort of situation where [GHC’s `Note`s](https://github.com/hasura/graphql-engine/blob/995e198d0af405a67415ef3d2a178a4c5bda1d54/server/STYLE.md#out-of-line-notes) really shine. It’s not immediately clear from reading this code why we’re fetching things from `pg_type` here, and we *could* figure it out by diligently following the control flow until we find the comment on the `CatalogCustomTypes` definition. But to make it easier on future readers, we could also add a `Note`:

```haskell
{- Note [Postgres scalars in custom types]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
It’s very convenient to be able to reference Postgres scalars in custom type
definitions. For example, we might have a type like this:

    type User {
      id: uuid!
      name: String
      location: geography
    }

The uuid and geography types are Postgres scalars, not separately-defined
GraphQL types. To support this, we have to take a few extra steps:

  1. The set of Postgres base types is not fixed; extensions like PostGIS add
     new ones, and users can even define their own. Therefore, we fetch the
     currently defined base types from the @pg_type@ system table as part of
     loading the metadata.

  2. It’s possible for a custom type definition to use a type that doesn’t
     appear elsewhere in the GraphQL schema, so we record which base types were
     referenced while validating the custom type definitions and make sure to
     include them in the generated schema explicitly. -}
```

You could put that somewhere in `Hasura.RQL.DDL.CustomTypes`, perhaps next to the `validateCustomTypeDefinitions` function, and then you could add a comment in `catalog_metadata.sql` that points to it:

```sql
      json_build_object(
        'custom_types',
         coalesce((select custom_types from hdb_catalog.hdb_custom_types), '{}'),
        'pg_scalars', -- see Note [Postgres scalars in custom types]
         coalesce((select json_agg(typname) from pg_catalog.pg_type where typtype = 'b'), '[]')
      ) as item
```

You could also include references to the `Note` in the comment attached to `CatalogCustomTypes` and in the body of the `validateCustomTypeDefinitions` function. That way you can just state all the tricky details in one spot, and you don’t have to restate yourself in several different places! 🙂